### DISCUSSION TITLE
Octaves support?

### DISCUSSION THREAD

**Discussion by jyoung4242** [jwagner/simplex-noise.js#56] [Q&A] [answered]
Is there a good way of using octaves with this library to smooth it out a bit?

**Comment by jwagner** [comment, marked as answer]
Hey Justin, I assume that with octaves you mean something like [FBM](https://thebookofshaders.com/13/). 

You can look at the synthwave demo I did for an example of that.
https://github.com/jwagner/simplex-noise-demo-synthwave/blob/13a5d96bada482b44a323c3b1ea6087102cb0ee2/fbm2d.ts
https://github.com/jwagner/simplex-noise-demo-synthwave/blob/13a5d96bada482b44a323c3b1ea6087102cb0ee2/renderer.ts#L8

I hope that helps. :)

**Reply by jyoung4242** [reply]
Fractal Brownian Movement?  let me read about it, you may very well be right, I'm kina a noob at this noise stuff

**Comment by jyoung4242** [comment]
Yup, this is what i was looking for!  Thx

**Comment by redblobgames** [comment]
I'm using jwagner's simplex-noise for this explanation of octaves: https://www.redblobgames.com/maps/terrain-from-noise/#octaves

**Reply by jwagner** [reply]
This is awesome! I should add a section with some tutorials - especially yours - to the README I think. :)

**Comment by jyoung4242** [comment]
@jwagner do you have a discord server or other forms of interaction that works better for you?
the noise pattern i THINK is working as designed, but maybe i'm using the wrong tool for the job

**Reply by jyoung4242** [reply]
follow up:, in the fbm method, i assume you can play with some of the parameters to change the nature of the noise' behavior?

**Reply by jyoung4242** [reply]
like maybe i should have targeted perlin vs simplex or something

**Reply by redblobgames** [reply]
Yes, fbm has lots of parameters you can play with. This isn't a perlin vs simplex thing; Perlin, Simplex, Midpoint Displacement all can be used with or without fbm.

The idea is that simplex/perlin noises give you a narrow "band" of frequencies. The audio analogy is that if you're listening to some music, it comes premixed. But if you want your bass to be high and treble to be low, you'd need to separate that music into low and middle and high frequencies. Then you can turn up the volume on the low frequencies (bass) and turn down the volume on the high frequencies (treble), and then add everything together again.

"Octaves" are a common way of dividing up the spectrum into bands. In each band you can decide what noise to put in. The standard fbm approach (which some libraries do for you) is to put more into bass and less into treble. This gives you standard fbm mountainous terrain. See the bottom of https://iquilezles.org/articles/fbm/  Some libraries give you this "standard mix" as a convenience.

For the standard mix, I use a helper function like this that takes an array of amplitudes. The "standard" mix is an amplitude array [1, 1/2, 1/4, 1/8, 1/16, etc.] where each amplitude is 1/2 as big as the previous (this is sometimes called "persistence"). But I've had better results by using my own mix. The standard octave approach is to multiply the frequency by 2 each time (this is called "lacunarity"). But you might want to experiment with that number too.

```js
function fbm_noise(noise, amplitudes, nx, ny) {
    let sum = 0, sumOfAmplitudes = 0;
    for (let octave = 0; octave < amplitudes.length; octave++) {
        let frequency = 1 << octave;
        sum += amplitudes[octave] * noise.noise2D(nx * frequency, ny * frequency, octave);
        sumOfAmplitudes += amplitudes[octave];
    }
    return sum / sumOfAmplitudes;
}
```

If you write that loop yourself instead of using a library function, you can change a *lot* more things. For example you might decide that the low frequency band should be generated using user-painted elevations, and middle and high frequency bands are generated using noise. (See [my project](https://www.redblobgames.com/maps/mapgen4/) for an example.) Or you might decide that for low elevations you want to make it flatter so you reduce amount of noise, and for high elevations you want to make it noisier so you increase the amount of noise. Or you might decide that you want to enhance the middle frequencies in deserts and enhance the low frequencies in forests. The library functions that give you fbm don't allow all of this, so I prefer to write the loop myself.
